File: src/main/java/supermetroid/MajorTracker.java
package supermetroid;

import supermetroid.trackable.Trackable;

import javax.imageio.ImageIO;
import javax.swing.*;
import java.awt.*;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public class MajorTracker extends JFrame {
  public boolean reset = false;
  private List<Trackable> trackables;

  public MajorTracker(Map<Point, Trackable> trackablesByLocation) throws Exception {
    setTitle("Super Metroid - Major Tracker");
    setResizable(false);
    setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
    addKeyListener(new KeyAdapter() {
      @Override
      public void keyPressed(KeyEvent e) {
        switch (e.getKeyChar()) {
          case 'r': reset = true; break;
          case 's': takeScreenshot(); break;
        }
      }
    });

    JPanel panel = new JPanel(new GridLayout(4,5,2,2));
    panel.setBackground(Color.BLACK);
    setContentPane(panel);
    trackables = new ArrayList<>(trackablesByLocation.values());

    var black = new ImageIcon(ImageIO.read(Objects.requireNonNull(MajorTracker.class.getResourceAsStream("/black.png"))));

    for (var row = 0; row < 4; row++) {
      for (var col = 0; col < 5; col++) {
        Point p;

        JLabel label =  new JLabel();
        panel.add(label);
        label.setSize(64,64);
        label.setIcon(black);

        if (trackablesByLocation.containsKey(p = new Point(col, row))) {
          trackablesByLocation.get(p).setLable(label);
        }
      }
    }

    pack();
  }

  public void updateState(State state) {
    trackables.forEach(trackable -> trackable.setIcon(state));
  }

  public void takeScreenshot() {
    BufferedImage image = new BufferedImage(getWidth(), getHeight(), BufferedImage.TYPE_INT_RGB);
    Graphics2D graphics2D = image.createGraphics();
    paint(graphics2D);
    try {
      ImageIO.write(image,"png", new File(System.getProperty("user.home") + "/Desktop/tracker.png"));
    } catch (IOException ex) {
      ex.printStackTrace();
    }
  }
}

File: src/main/java/supermetroid/MajorTracker.kt
package supermetroid

import supermetroid.trackable.Trackable
import java.awt.Color
import java.awt.GridLayout
import java.awt.Point
import java.awt.event.KeyAdapter
import java.awt.event.KeyEvent
import java.awt.image.BufferedImage
import java.io.File
import java.io.IOException
import java.util.*
import java.util.function.Consumer
import javax.imageio.ImageIO
import javax.swing.JFrame
import javax.swing.JLabel
import javax.swing.JPanel

class MajorTracker(trackablesByLocation: Map<Point, Trackable>) : JFrame() {
    var reset: Boolean = false
    val trackables: List<Trackable>

    init {
        title = "Super Metroid - Major Tracker"
        isResizable = false
        defaultCloseOperation = EXIT_ON_CLOSE
        addKeyListener(object : KeyAdapter() {
            override fun keyPressed(e: KeyEvent) {
                when (e.keyChar) {
                    'r' -> reset = true
                    's' -> takeScreenshot()
                }
            }
        })

        val panel = JPanel(GridLayout(4, 5, 2, 2))
        panel.background = Color.BLACK
        contentPane = panel
        trackables = ArrayList(trackablesByLocation.values)

        val black = readResourceToIcon("/black.png")

        for (row in 0..3) {
            for (col in 0..4) {
                var p: Point

                val label = JLabel()
                panel.add(label)
                label.setSize(64, 64)
                label.icon = black

                if (trackablesByLocation.containsKey(Point(col, row).also { p = it })) {
                    trackablesByLocation[p]!!.initLabel(label)
                }
            }
        }

        pack()
    }

    fun updateState(state: State) {
        trackables.forEach(Consumer { trackable: Trackable -> trackable.setIcon(state) })
    }

    fun takeScreenshot() {
        val image = BufferedImage(width, height, BufferedImage.TYPE_INT_RGB)
        val graphics2D = image.createGraphics()
        paint(graphics2D)
        try {
            ImageIO.write(image, "png", File(System.getProperty("user.home") + "/Desktop/tracker.png"))
        } catch (ex: IOException) {
            ex.printStackTrace()
        }
    }
}
